File: backend/checker.js
const net = require('net');

/**
 * Probe a single proxy host:port via TCP socket to check if open and measure round-trip latency (ms)
 */
function probeProxy(host, port, timeoutMs = 2000) {
  return new Promise((resolve) => {
    const startTime = Date.now();
    const socket = new net.Socket();

    let resolved = false;
    const cleanup = () => {
      socket.removeAllListeners();
      socket.destroy();
    };

    socket.setTimeout(timeoutMs);

    socket.connect(port, host, () => {
      if (!resolved) {
        resolved = true;
        const latency = Date.now() - startTime;
        cleanup();
        resolve({ alive: true, latency });
      }
    });

    socket.on('timeout', () => {
      if (!resolved) {
        resolved = true;
        cleanup();
        resolve({ alive: false, latency: null, reason: 'timeout' });
      }
    });

    socket.on('error', (err) => {
      if (!resolved) {
        resolved = true;
        cleanup();
        resolve({ alive: false, latency: null, reason: err.code || 'error' });
      }
    });
  });
}

/**
 * Check a batch of proxies with concurrency limit
 */
async function batchCheckProxies(proxies, concurrency = 50, timeoutMs = 2000, onProgress = null) {
  let index = 0;
  let active = 0;
  let completed = 0;
  const total = proxies.length;

  return new Promise((resolve) => {
    if (total === 0) return resolve();

    function next() {
      while (active < concurrency && index < total) {
        const currentIndex = index++;
        const proxy = proxies[currentIndex];
        active++;

        probeProxy(proxy.host, proxy.port, timeoutMs).then((res) => {
          proxy.alive = res.alive;
          if (res.alive) {
            proxy.latency = res.latency;
          }
          proxy.lastChecked = Date.now();
          active--;
          completed++;

          if (onProgress && completed % 25 === 0) {
            onProgress(completed, total);
          }

          if (completed === total) {
            resolve();
          } else {
            next();
          }
        });
      }
    }

    next();
  });
}

module.exports = {
  probeProxy,
  batchCheckProxies
};

File: backend/src/checker.rs
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::net::TcpStream;
use tokio::sync::Semaphore;
use tokio::time::timeout;
use tracing::info;
use crate::models::Proxy;

/// Probe a single proxy's host:port over TCP
pub async fn probe_proxy(host: &str, port: u16, timeout_duration: Duration) -> (bool, Option<u32>) {
    let target = format!("{}:{}", host, port);
    let start = Instant::now();

    match timeout(timeout_duration, TcpStream::connect(&target)).await {
        Ok(Ok(_stream)) => {
            let elapsed = start.elapsed().as_millis() as u32;
            (true, Some(elapsed))
        }
        _ => (false, None),
    }
}

/// Concurrently probe a slice of proxies up to limit using a semaphore
pub async fn probe_proxies_batch(proxies: &mut [Proxy], concurrency: usize, timeout_duration: Duration) {
    let semaphore = Arc::new(Semaphore::new(concurrency));
    let mut handles = Vec::new();

    let now_ts = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64;

    for (idx, p) in proxies.iter().enumerate() {
        let sem = semaphore.clone();
        let host = p.host.clone();
        let port = p.port;

        handles.push(tokio::spawn(async move {
            let _permit = sem.acquire().await.unwrap();
            let (alive, latency) = probe_proxy(&host, port, timeout_duration).await;
            (idx, alive, latency)
        }));
    }

    let mut completed = 0;
    let total = handles.len();

    for handle in handles {
        if let Ok((idx, alive, latency)) = handle.await {
            proxies[idx].alive = Some(alive);
            if alive {
                proxies[idx].latency = latency;
            }
            proxies[idx].last_checked = Some(now_ts);
            completed += 1;
            if completed % 100 == 0 || completed == total {
                info!("[Checker] Latency probe progress: {}/{}", completed, total);
            }
        }
    }
}
